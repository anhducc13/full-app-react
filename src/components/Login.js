import React, { useState } from 'react';
import { setGlobal } from 'reactn';
import Swal from 'sweetalert2'
import { LoginForm } from 'views/LoginForm'
import {
  validateLength,
  haveSpace,
  haveUpperCharacter,
  haveLowerCharacter,
  haveNumberCharacter,
  haveSpecialCharacter
} from 'helpers/validators'
import { authService } from 'services'

export const Login = (props) => {
  const initTextField = {
    value: '',
    error: '',
    valid: false,
  }

  const [usernameField, setUsernameField] = useState(initTextField);
  const [passwordField, setPasswordField] = useState(initTextField);

  const onChangeUsername = (e) => {
    let { value, error, valid } = usernameField;
    value = e.target.value;
    if (
      !haveLowerCharacter(value.toLowerCase())
      || !haveNumberCharacter(value)
      || haveSpecialCharacter(value)
    ) {
      error = 'Tên đăng nhập bao gồm chữ cái và số';
      valid = false;
    } else if (haveSpace(value)) {
      error = 'Tên đăng nhập không được chứa khoảng trắng';
      valid = false;
    } else if (!validateLength(value, { min: 6, max: 128 })) {
      error = 'Tên đăng nhập có độ dài từ 6 đến 128 ký tự';
      valid = false;
    } else {
      error = '';
      valid = true;
    }
    setUsernameField({
      value: value,
      error: error,
      valid: valid,
    });
  }

  const onChangePassword = (e) => {
    let { value, error, valid } = passwordField;
    value = e.target.value;
    if (
      !haveLowerCharacter(value)
      || !haveUpperCharacter(value)
      || !haveNumberCharacter(value)
      || haveSpecialCharacter(value)
    ) {
      error = 'Mật khẩu bao gồm chữ thường, chữ hoa và số';
      valid = false;
    } else if (haveSpace(value)) {
      error = 'Mật khẩu không được chứa khoảng trắng';
      valid = false;
    } else if (!validateLength(value, { min: 8, max: 128 })) {
      error = 'Mật khẩu có độ dài từ 8 đến 128 ký tự';
      valid = false;
    } else {
      error = '';
      valid = true;
    }
    setPasswordField({
      value: value,
      error: error,
      valid: valid,
    });
  }

  const submitForm = (e) => {
    e.preventDefault()
    const params = {
      "username": usernameField.value,
      "password": passwordField.value
    }
    setGlobal({
      loading: true
    });
    authService.login(params)
      .then(res => {
        setGlobal({
          loading: false
        });
        if (res.status === 200) {
          const userInfoLogin = res.data;
          localStorage.setItem('userInfoLogin', JSON.stringify(userInfoLogin));
          Swal.fire(
            'Thành công!',
            `Xin chào ${userInfoLogin['username']} quay trở lại`,
            'success'
          )
            .then(() => {
              props.history.push('/');
            })
        }
      })
      .catch(err => {
        setGlobal({
          loading: false
        });
        if (err.response.status && err.response.status === 404) {
          Swal.fire(
            'Có lỗi xảy ra!',
            'Tài khoản hoặc mật khẩu không chính xác',
            'error'
          )
        } else if (err.response.status && err.response.status === 400) {
          Swal.fire(
            'Có lỗi xảy ra!',
            'Dữ liệu đầu vào có vấn đề',
            'error'
          )
        } else {
          props.history.push('/error');
        }
      })
      .finally(() => {
        setGlobal({
          loading: false
        });
      })
  }

  return (
    <div>
      <LoginForm
        textField={{
          usernameField: usernameField,
          passwordField: passwordField
        }}
        handleChange={{
          onChangeUsername: onChangeUsername,
          onChangePassword: onChangePassword,
        }}
        handleSubmit={submitForm}
      />
    </div>
  )
}
