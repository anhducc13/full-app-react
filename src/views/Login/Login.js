import React from 'react';
import { LoginForm } from 'components/LoginForm';
import { Auth } from 'helpers/auth';
import { Redirect } from 'react-router-dom';

export const Login = (props) => {
  return (
    <div>
      {Auth.isAuthenticated() ? (
        <Redirect to="/home" />
      ) : (
          <LoginForm {...props}/>
        )}
    </div>
  )
}
