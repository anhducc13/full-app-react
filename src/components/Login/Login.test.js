import React from 'react';
import { shallow, mount } from 'enzyme';

import { Login } from './Login';

describe('Login', () => {
  it('should render correctly', () => {
    const component = shallow(<Login />);
    expect(component).toMatchSnapshot();
  });
});

describe('Login tests included components', () => {

  it('renders login form', () => {
    const wrapper = shallow(<Login />);
    expect(wrapper.find('LoginForm')).toBeDefined();
  });
  it('redirect to home', () => {
    const wrapper = shallow(<Login />);
    expect(wrapper.find('Redirect')).toBeDefined();
  });
});
