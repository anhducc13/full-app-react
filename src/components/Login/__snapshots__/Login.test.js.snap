// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should render correctly 1`] = `
<div>
  <LoginForm
    handleChange={
      Object {
        "onChangePassword": [Function],
        "onChangeUsername": [Function],
      }
    }
    handleSubmit={[Function]}
    textField={
      Object {
        "passwordField": Object {
          "error": "",
          "valid": false,
          "value": "",
        },
        "usernameField": Object {
          "error": "",
          "valid": false,
          "value": "",
        },
      }
    }
  />
</div>
`;
